### PR TITLE
Diagnostics

### PR DESCRIPTION
**Pull request by mrocklin**
This replicates the dask `ProgressBar` within distributed.  It works:

### API

See https://gist.github.com/mrocklin/ebf84803c9e86966004f

```python
>>> progress(futures)
[########################################] | 100% Completed |  5.3s
```

### Diagnostic outline

Currently diagnostics register themselves with the `Scheduler` and get updated on certain events.  Currently we call diagnostic methods on `update_graph`, `task_erred` and `task_finished` events, though we can add more easily.  Diagnostics can do whatever they want to when called.  Currently I either spin up a thread and update internal values or I insert a `PeriodicCallback` into Jupyter's IOLoop, but we could communicate over sockets or whatnot instead.

We could consider other approaches, like periodically running an arbitrary function over the scheduler state.  This might be useful for other visualizations.

cc @freeman-lab @broxtronix

(thanks to @jcrist from whose dask code I stole avidly.  Technical feedback welcome here Jim.  I'm probably doing dumb things. This was made in a rush.)

https://github.com/blaze/distributed/issues/21

### CODE REVIEW

File: distributed/diagnostics.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        self.keys = set()
+
+        self._running = False
+
+        if scheduler is None and default_executor():
+            scheduler = default_executor().scheduler
+
+        if scheduler:
+            self.scheduler = scheduler
+            scheduler.add_diagnostic(self)
```
Comment (jcrist) [inline distributed/diagnostics.py]: The `scheduler is None and not default_executor()` case is unhandled. Not sure if that's important.

**Comment by jcrist** [comment]
Overall this looks sane to me.

File: distributed/diagnostics.py (hunk 1 above)
Comment (mrocklin) [inline distributed/diagnostics.py]: Resolved with an error message

**Comment by mrocklin** [comment]
Planning to merge this soon in order to get some feedback on it.  I expect this code to churn a fair amount in the next few days.

**Comment by broxtronix** [comment]
Very excited to have a progress bar!  The `Dask` progress bar works great so this will largely address the immediate need to keep track of the progress of computations, even if it is blocking for now.  

However, I took this out for a spin, and it didn't quite work as intended.  It doesn't seem to flush the output properly in the iPython notebook, so I only see the final update when the computation is complete.  I tried the `dask.diagnostics.ProgressBar()` when using the `dask.threaded` and `dask.multiprocessing` execution engines, and it seems to work and update as intended, so something must be funky with the `multiprocessing.diagnostics.ProgressBar()`.  

The code looks reasonable enough... it seems to `flush()` the output when it should and I cannot see any other bug in the threading code.  I'm not sure why this doesn't update properly in the iPython notebook.  Does it work there for you?

**Comment by mrocklin** [comment]
I've only been testing in the console.  Will test things out now in the notebook.

**Comment by mrocklin** [comment]
@jcrist I believe that you ran into this with the dask `ProgressBar`.  Can you shed any light here?

**Comment by jcrist** [comment]
Only calls to `flush` in the main thread will result in the progressbar updating in the notebook. Without looking at how distributed works, I assume this isn't happening (see commit in dask here for more info: https://github.com/jcrist/dask/commit/79ae559e604590ca0469a38a5e87d5b1e362e2bf).

**Comment by freeman-lab** [comment]
@mrocklin just tried this out in the console, works nicely! The API for both `wait` and the context manager feel good, one nit: when calling `wait(L, show_progress=True)` on a list of `Futures` that have already completed, the bar displays with `0%` rather than `100%` (see below) and a short completion time, should probably show a full bar and 100%? Can look more at the code in a sec to spot where this happens but you'll probably know right away.

```python
In [1]: wait(L, show_progress=True);
[########################################] | 100% Completed | 15.1s

In [2]: wait(L, show_progress=True);
[                                        ] | 0% Completed |  0.2s
```

**Comment by mrocklin** [comment]
Oh, good find.  Fixed in https://github.com/blaze/distributed/commit/5165ebe65f94e2798f89ab86103e74a7c068bbf2

**Comment by mrocklin** [comment]
> Only calls to flush in the main thread will result in the progressbar updating in the notebook. Without looking at how distributed works, I assume this isn't happening (see commit in dask here for more info: jcrist/dask@79ae559).

Hrm, that is a bit of a pain.  The scheduler itself is running in a background thread (unlike dask, where it's in the main thread).  It might be worth skipping this problem and going into something a bit more fancy like what was recommended in this stackoverflow post: http://stackoverflow.com/questions/34033652/create-jupyter-notebook-object-that-changes-after-the-cell-has-run

**Comment by mrocklin** [comment]
OK, this now has a fully asynchronous IPython widget progressbar.  This is a bit hard to show over static web, but if you run the following notebook under this branch then things should be dandy.

https://gist.github.com/mrocklin/ebf84803c9e86966004f

I've updated the header comment a bit.  Notably we no longer have context manager abilities.  You must have a `Future` object to use the progress bar.  I'll do a bit of work later to help dask collections integrate with this.

I don't yet have reasonable tests for the IPython stuff.  I'll need to find some setup-teardown solution for ipython kernels (any suggestions anyone)?

**Comment by mrocklin** [comment]
More untested (but neat) work.  https://gist.github.com/mrocklin/e14a3752bc0e7df33efd

This adds a multi-progress-bar that tracks the progress of all functions in the computation separately, similar to this example:

```
all:      [|||||||||------------------] 30%
load:     [|||||||||||||||||||||||----] 90%
compute:  [||||-----------------------] 10%
store:    [||-------------------------]  5%
other:    [|||||||||||||||------------] 50%
```

**Comment by mrocklin** [comment]
Now with tests!

**Comment by mrocklin** [comment]
So, some general thoughts:

It's going to be important to separate the collection/aggregation, and presentation of diagnostic information.  

The collection/aggregation should probably occur as a plugin to the scheduler (for which we seem to have a decent thing going with the abstract `Diagnostic` class (perhaps soon to be named `Plugin`).  This is just an object with callbacks for various events.  We should decouple this from any kind of visualization because we're likely to want several different kinds of visualizations.

We've seen value to this decoupling here with the Text and Widget progressbars, but we can expect further UI outputs like Bokeh dashboards and perhaps just straight JS code.  Also, when we further separate the Scheduler and Executor to live on different machines our `Diagnostic/Plugin` objects will live on one machine and communicate via sockets to visualizations on another.

**Comment by mrocklin** [comment]
My plan is to clean up this PR, merge, and then start thinking about collecting and presenting CPU/Memory information from the Workers.  Hopefully having two use cases will help to nail down a more general diagnostic abstraction.

**Comment by mrocklin** [comment]
Merging tomorrow morning if no comments.  This undoubtedly still needs some work.  I'm planning to merge and work on some related issues for a bit though and then come back to the design of this.

**Pull request merged**